# Data Wrangling

### Import Data

In [1]:
# Import Necessary Packages
import pandas as pd
import numpy as np

In [2]:
# Import Data from CSV file
df = pd.read_csv(r'C:\Users\bronc\Downloads\Capstone 3\sales_data_sample.csv', encoding='Latin-1')

In [3]:
df.head()

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

         ORDERDATE   STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0   2/24/2003 0:00  Shipped       1         2     2003  ...   
1    5/7/2003 0:00  Shipped       2         5     2003  ...   
2    7/1/2003 0:00  Shipped       3         7     2003  ...   
3   8/25/2003 0:00  Shipped       3         8     2003  ...   
4  10/10/2003 0:00  Shipped       4        10     2003  ...   

                    ADDRESSLINE1  ADDRESSLINE2           CITY STATE  \
0        897 Long Airport Avenue           NaN            NYC    NY   
1             59 rue de l'Abbaye           NaN          Reims   NaN   
2  27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3             78934 Hillside Dr.           NaN       Pasadena    CA   
4                7734 Strong St.           NaN  San Francisco    CA   

  POSTALCODE COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0      10022     USA       NaN              Yu             Kwai    Small  
1      51100  France      EMEA         Henriot             Paul    Small  
2      75508  France      EMEA        Da Cunha           Daniel   Medium  
3      90003     USA       NaN           Young            Julie   Medium  
4        NaN     USA       NaN           Brown            Julie   Medium  

[5 rows x 25 columns]

### Data Definition

In [4]:
# Examine info about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ORDERNUMBER       2823 non-null   int64  
 1   QUANTITYORDERED   2823 non-null   int64  
 2   PRICEEACH         2823 non-null   float64
 3   ORDERLINENUMBER   2823 non-null   int64  
 4   SALES             2823 non-null   float64
 5   ORDERDATE         2823 non-null   object 
 6   STATUS            2823 non-null   object 
 7   QTR_ID            2823 non-null   int64  
 8   MONTH_ID          2823 non-null   int64  
 9   YEAR_ID           2823 non-null   int64  
 10  PRODUCTLINE       2823 non-null   object 
 11  MSRP              2823 non-null   int64  
 12  PRODUCTCODE       2823 non-null   object 
 13  CUSTOMERNAME      2823 non-null   object 
 14  PHONE             2823 non-null   object 
 15  ADDRESSLINE1      2823 non-null   object 
 16  ADDRESSLINE2      302 non-null    object 


From this list I can see that some columns have null values. I will take a look and see how to handle these in the data cleaning step

In [5]:
df.describe()

ORDERNUMBER  QUANTITYORDERED    PRICEEACH  ORDERLINENUMBER  \
count   2823.000000      2823.000000  2823.000000      2823.000000   
mean   10258.725115        35.092809    83.658544         6.466171   
std       92.085478         9.741443    20.174277         4.225841   
min    10100.000000         6.000000    26.880000         1.000000   
25%    10180.000000        27.000000    68.860000         3.000000   
50%    10262.000000        35.000000    95.700000         6.000000   
75%    10333.500000        43.000000   100.000000         9.000000   
max    10425.000000        97.000000   100.000000        18.000000   

              SALES       QTR_ID     MONTH_ID     YEAR_ID         MSRP  
count   2823.000000  2823.000000  2823.000000  2823.00000  2823.000000  
mean    3553.889072     2.717676     7.092455  2003.81509   100.715551  
std     1841.865106     1.203878     3.656633     0.69967    40.187912  
min      482.130000     1.000000     1.000000  2003.00000    33.000000  
25%     2203.430000     2.000000     4.000000  2003.00000    68.000000  
50%     3184.800000     3.000000     8.000000  2004.00000    99.000000  
75%     4508.000000     4.000000    11.000000  2004.00000   124.000000  
max    14082.800000     4.000000    12.000000  2005.00000   214.000000

In [6]:
# Next I'll look at the number of unique values per column
df.nunique()

ORDERNUMBER          307
QUANTITYORDERED       58
PRICEEACH           1016
ORDERLINENUMBER       18
SALES               2763
ORDERDATE            252
STATUS                 6
QTR_ID                 4
MONTH_ID              12
YEAR_ID                3
PRODUCTLINE            7
MSRP                  80
PRODUCTCODE          109
CUSTOMERNAME          92
PHONE                 91
ADDRESSLINE1          92
ADDRESSLINE2           9
CITY                  73
STATE                 16
POSTALCODE            73
COUNTRY               19
TERRITORY              3
CONTACTLASTNAME       77
CONTACTFIRSTNAME      72
DEALSIZE               3
dtype: int64

### Data Cleaning

Before looking into the individual columns let's fix the column names so they look proper

In [7]:
df.columns = ['Order_Number', 'Quantity_Ordered', 'Price_Each', 'Order_Line_Number', 'Sales', 'Order_Date', 'Status', 'QTR_ID', 'Month_ID', 'Year_ID', 'Product_Line', 'MSRP', 'Product_Code', 'Customer_Name', 'Phone', 'Address_Line_1', 'Address_Line_2', 'City', 'State', 'Postal_Code', 'Country', 'Territory', 'Contact_Last_Name', 'Contact_First_Name', 'Deal_Size']

In [8]:
# Verify renamed columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Order_Number        2823 non-null   int64  
 1   Quantity_Ordered    2823 non-null   int64  
 2   Price_Each          2823 non-null   float64
 3   Order_Line_Number   2823 non-null   int64  
 4   Sales               2823 non-null   float64
 5   Order_Date          2823 non-null   object 
 6   Status              2823 non-null   object 
 7   QTR_ID              2823 non-null   int64  
 8   Month_ID            2823 non-null   int64  
 9   Year_ID             2823 non-null   int64  
 10  Product_Line        2823 non-null   object 
 11  MSRP                2823 non-null   int64  
 12  Product_Code        2823 non-null   object 
 13  Customer_Name       2823 non-null   object 
 14  Phone               2823 non-null   object 
 15  Address_Line_1      2823 non-null   object 
 16  Addres

For this step I'll be going through the columns where I want to examine the data to ensure its clean or where there are nulls to decide how they should be handled

In [9]:
#Ensure Sales is calculated correctly
df.sample(10, random_state = 123)

Order_Number  Quantity_Ordered  Price_Each  Order_Line_Number    Sales  \
1511         10343                27       36.21                  6   977.67   
2766         10386                44       86.40                 15  3801.60   
558          10178                42      100.00                  4  6490.68   
591          10288                32      100.00                  5  5875.20   
1590         10105                43      100.00                  9  6341.21   
2087         10280                21       78.89                  6  1656.69   
1857         10279                33       71.06                  4  2344.98   
2256         10176                22       64.00                  6  1408.00   
1087         10383                26      100.00                 12  3340.48   
1420         10182                47       74.22                 16  3488.34   

           Order_Date    Status  QTR_ID  Month_ID  Year_ID  ...  \
1511  11/24/2004 0:00   Shipped       4        11     2004  ...   
2766    3/1/2005 0:00  Resolved       1         3     2005  ...   
558    11/8/2003 0:00   Shipped       4        11     2003  ...   
591     9/1/2004 0:00   Shipped       3         9     2004  ...   
1590   2/11/2003 0:00   Shipped       1         2     2003  ...   
2087   8/17/2004 0:00   Shipped       3         8     2004  ...   
1857    8/9/2004 0:00   Shipped       3         8     2004  ...   
2256   11/6/2003 0:00   Shipped       4        11     2003  ...   
1087   2/22/2005 0:00   Shipped       1         2     2005  ...   
1420  11/12/2003 0:00   Shipped       4        11     2003  ...   

                               Address_Line_1  Address_Line_2           City  \
1511                       59 rue de l'Abbaye             NaN          Reims   
2766                       C/ Moralzarzal, 86             NaN         Madrid   
558                     1 rue Alsace-Lorraine             NaN       Toulouse   
591   Village Close - 106 Linden Road Sandown       2nd Floor      Singapore   
1590                             Vinb'ltet 34             NaN      Kobenhavn   
2087                      Via Monte Bianco 34             NaN         Torino   
1857                       C/ Moralzarzal, 86             NaN         Madrid   
2256                   Strada Provinciale 124             NaN  Reggio Emilia   
1087                       C/ Moralzarzal, 86             NaN         Madrid   
1420                          5677 Strong St.             NaN     San Rafael   

     State Postal_Code    Country Territory Contact_Last_Name  \
1511   NaN       51100     France      EMEA           Henriot   
2766   NaN       28034      Spain      EMEA            Freyre   
558    NaN       31000     France      EMEA            Roulet   
591    NaN       69045  Singapore      APAC         Victorino   
1590   NaN        1734    Denmark      EMEA          Petersen   
2087   NaN       10100      Italy      EMEA           Accorti   
1857   NaN       28034      Spain      EMEA            Freyre   
2256   NaN       42100      Italy      EMEA            Moroni   
1087   NaN       28034      Spain      EMEA            Freyre   
1420    CA       97562        USA       NaN            Nelson   

     Contact_First_Name Deal_Size  
1511               Paul     Small  
2766              Diego    Medium  
558             Annette    Medium  
591               Wendy    Medium  
1590              Jytte    Medium  
2087              Paolo     Small  
1857              Diego     Small  
2256           Maurizio     Small  
1087              Diego    Medium  
1420            Valarie    Medium  

[10 rows x 25 columns]

In [10]:
print(27 * 36.21)
print(44 * 86.40)
print(42 * 100)
print(32 * 100)
print(43 * 100)
print(21 * 78.89)
print(33 * 71.06)
print(22 * 64)
print(26 * 100)
print(47 * 74.22)

977.6700000000001
3801.6000000000004
4200
3200
4300
1656.69
2344.98
1408
2600
3488.34


Of these 10 rows 4 have values that don't seem to be quantity ordered times price each. I will explore some other columns to see if there's a reason for this in the data. If not I will likely make my own columns for revenue

In [11]:
# Change Order_Date to a datetime object
df['Order_Date'] = pd.to_datetime(df['Order_Date'], errors = 'coerce')

In [12]:
# Check for new nulls
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Order_Number        2823 non-null   int64         
 1   Quantity_Ordered    2823 non-null   int64         
 2   Price_Each          2823 non-null   float64       
 3   Order_Line_Number   2823 non-null   int64         
 4   Sales               2823 non-null   float64       
 5   Order_Date          2823 non-null   datetime64[ns]
 6   Status              2823 non-null   object        
 7   QTR_ID              2823 non-null   int64         
 8   Month_ID            2823 non-null   int64         
 9   Year_ID             2823 non-null   int64         
 10  Product_Line        2823 non-null   object        
 11  MSRP                2823 non-null   int64         
 12  Product_Code        2823 non-null   object        
 13  Customer_Name       2823 non-null   object      

No new nulls and the Dtype changed!

In [13]:
# Check different values for Status
df['Status'].unique()

array(['Shipped', 'Disputed', 'In Process', 'Cancelled', 'On Hold',
       'Resolved'], dtype=object)

In [14]:
# Inspect the value counts of each of these statuses
df['Status'].value_counts()

Shipped       2617
Cancelled       60
Resolved        47
On Hold         44
In Process      41
Disputed        14
Name: Status, dtype: int64

In [15]:
# Check percentage of shipped status
2617/2823

0.9270279844137442

Shipped items account for 92.7% of the data. While resolved is also a likely sale I don't have any way of being sure of that. For that reason I will be dropping all other statuses other than Shipped as this doesn't account for a big percentage of the data and I can safely assume that these were successful sales

In [16]:
# Keep only Shipped statuses
df = df[df.Status == 'Shipped']

In [17]:
# Check that only 2617 rows remain
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2617 entries, 0 to 2821
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Order_Number        2617 non-null   int64         
 1   Quantity_Ordered    2617 non-null   int64         
 2   Price_Each          2617 non-null   float64       
 3   Order_Line_Number   2617 non-null   int64         
 4   Sales               2617 non-null   float64       
 5   Order_Date          2617 non-null   datetime64[ns]
 6   Status              2617 non-null   object        
 7   QTR_ID              2617 non-null   int64         
 8   Month_ID            2617 non-null   int64         
 9   Year_ID             2617 non-null   int64         
 10  Product_Line        2617 non-null   object        
 11  MSRP                2617 non-null   int64         
 12  Product_Code        2617 non-null   object        
 13  Customer_Name       2617 non-null   object      

In [18]:
# Verify that Shipped is only value in Status
df['Status'].value_counts()

Shipped    2617
Name: Status, dtype: int64

In [19]:
# Check data in Product_Line
df['Product_Line'].value_counts()

Classic Cars        914
Vintage Cars        557
Motorcycles         324
Trucks and Buses    281
Planes              271
Ships               195
Trains               75
Name: Product_Line, dtype: int64

This looks fine. We are looking across all these types of products to find the highest revenue generating individual product code so none of these will be dropped

Address_Line_1, Address_Line_2, State, Postal_Code and Territory are not needed because City and Country will give us sufficient data without nulls. Likewise we aren't looking at individual people so Phone, Contact_Last_Name and Contact_First_Name can be dropped since we have Customer_Name

In [20]:
# Remove unnecessary columns
df.drop(columns=['Phone', 'Address_Line_1', 'Address_Line_2', 'State', 'Postal_Code', 'Territory', 'Contact_Last_Name', 'Contact_First_Name' ], inplace = True)

In [21]:
# Verify columns were dropped
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2617 entries, 0 to 2821
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Order_Number       2617 non-null   int64         
 1   Quantity_Ordered   2617 non-null   int64         
 2   Price_Each         2617 non-null   float64       
 3   Order_Line_Number  2617 non-null   int64         
 4   Sales              2617 non-null   float64       
 5   Order_Date         2617 non-null   datetime64[ns]
 6   Status             2617 non-null   object        
 7   QTR_ID             2617 non-null   int64         
 8   Month_ID           2617 non-null   int64         
 9   Year_ID            2617 non-null   int64         
 10  Product_Line       2617 non-null   object        
 11  MSRP               2617 non-null   int64         
 12  Product_Code       2617 non-null   object        
 13  Customer_Name      2617 non-null   object        
 14  City    

Now all the columns have no null values. Next let's look at Sales again to see if we can figure out how it's calculated

In [22]:
df.sample(10, random_state = 123)

Order_Number  Quantity_Ordered  Price_Each  Order_Line_Number    Sales  \
2190         10340                55      100.00                  2  6482.85   
2256         10176                22       64.00                  6  1408.00   
1747         10143                27       60.97                  8  1646.19   
330          10282                24      100.00                  4  3778.80   
1562         10409                61       29.54                  1  1801.94   
1162         10363                24      100.00                 11  4142.64   
836          10173                22      100.00                  7  3452.68   
524          10381                25       52.83                  9  1320.75   
2271         10359                22      100.00                  1  4301.22   
200          10282                41      100.00                  5  7071.27   

     Order_Date   Status  QTR_ID  Month_ID  Year_ID  Product_Line  MSRP  \
2190 2004-11-24  Shipped       4        11     2004   Motorcycles    99   
2256 2003-11-06  Shipped       4        11     2003        Trains    62   
1747 2003-08-10  Shipped       3         8     2003        Planes    68   
330  2004-08-20  Shipped       3         8     2004  Classic Cars   173   
1562 2005-04-23  Shipped       2         4     2005  Vintage Cars    33   
1162 2005-01-06  Shipped       1         1     2005  Classic Cars   146   
836  2003-11-05  Shipped       4        11     2003  Vintage Cars   168   
524  2005-02-17  Shipped       1         2     2005  Vintage Cars    53   
2271 2004-12-15  Shipped       4        12     2004        Trains    62   
200  2004-08-20  Shipped       3         8     2004  Classic Cars   207   

     Product_Code                 Customer_Name           City    Country  \
2190     S32_1374            Enaco Distributors      Barcelona      Spain   
2256     S32_3207           L'ordine Souveniers  Reggio Emilia      Italy   
1747     S24_2841           Mini Creations Ltd.    New Bedford        USA   
330      S12_3891  Mini Gifts Distributors Ltd.     San Rafael        USA   
1562     S24_1937              Handji Gifts& Co      Singapore  Singapore   
1162     S18_3482           Suominen Souveniers          Espoo    Finland   
836      S18_2795                 Rovelli Gifts        Bergamo      Italy   
524      S18_1367      Corporate Gift Ideas Co.  San Francisco        USA   
2271     S32_3207            Reims Collectables          Reims     France   
200      S12_1108  Mini Gifts Distributors Ltd.     San Rafael        USA   

     Deal_Size  
2190    Medium  
2256     Small  
1747     Small  
330     Medium  
1562     Small  
1162    Medium  
836     Medium  
524      Small  
2271    Medium  
200      Large

In [23]:
print(55 * 100)
print(22 * 64)
print(27 * 60.97)
print(24 * 100)
print(61 * 29.54)
print(24 * 100)
print(22 * 100)
print(25 * 52.83)
print(22 * 100)
print(41 * 100)

5500
1408
1646.19
2400
1801.94
2400
2200
1320.75
2200
4100


Looking over this and comparing with the previous sample it appears that the Price_Each column is actually the issue. Anytime Price_Each = 100.00 the calculation is too small. This is likely because 100.00 was the maximum value being input. This can be fixed by recalculating this column as Sales/Quantity_Ordered

In [24]:
df['Price_Each'] = df['Sales']/df['Quantity_Ordered']

In [25]:
#Verify that the columns are properly calculated
df.head()

Order_Number  Quantity_Ordered  Price_Each  Order_Line_Number    Sales  \
0         10107                30       95.70                  2  2871.00   
1         10121                34       81.35                  5  2765.90   
2         10134                41       94.74                  2  3884.34   
3         10145                45       83.26                  6  3746.70   
4         10159                49      106.23                 14  5205.27   

  Order_Date   Status  QTR_ID  Month_ID  Year_ID Product_Line  MSRP  \
0 2003-02-24  Shipped       1         2     2003  Motorcycles    95   
1 2003-05-07  Shipped       2         5     2003  Motorcycles    95   
2 2003-07-01  Shipped       3         7     2003  Motorcycles    95   
3 2003-08-25  Shipped       3         8     2003  Motorcycles    95   
4 2003-10-10  Shipped       4        10     2003  Motorcycles    95   

  Product_Code             Customer_Name           City Country Deal_Size  
0     S10_1678         Land of Toys Inc.            NYC     USA     Small  
1     S10_1678        Reims Collectables          Reims  France     Small  
2     S10_1678           Lyon Souveniers          Paris  France    Medium  
3     S10_1678         Toys4GrownUps.com       Pasadena     USA    Medium  
4     S10_1678  Corporate Gift Ideas Co.  San Francisco     USA    Medium

In [26]:
49 * 106.23

5205.27

Here we have a column with Price_Each over 100 and the calculation has been fixed

In [27]:
# Save cleaned file for future work
df.to_csv(r'C:\Users\bronc\Downloads\Capstone 3\sales_data_sample(clean).csv')